tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp

```diff
@@ -0,0 +1,142 @@
+};
+
+} // namespace Impl
+
+namespace Impl {
+
+// Note: phrase it simply and the same as in specification ("has unique layout")
+template <typename Layout,
+          std::experimental::extents<>::size_type numRows,
+          std::experimental::extents<>::size_type numCols>
+inline constexpr bool is_unique_layout_v = Layout::template mapping<
+    std::experimental::extents<numRows, numCols> >::is_always_unique();
+
+template <typename Layout>
+struct is_layout_blas_packed: public std::false_type {};
+
+template <typename Triangle, typename StorageOrder>
+struct is_layout_blas_packed<
+  std::experimental::linalg::layout_blas_packed<Triangle, StorageOrder>>:
+    public std::true_type {};
+
+template <typename Layout>
+inline constexpr bool is_layout_blas_packed_v = is_layout_blas_packed<Layout>::value;
+
+// Note: will only signal failure for layout_blas_packed with diferent triangle
+template <typename Layout, typename Triangle>
+struct triangle_layout_mismatch: public std::false_type {};
+
+template <typename StorageOrder, typename Triangle1, typename Triangle2>
+struct triangle_layout_mismatch<
+  std::experimental::linalg::layout_blas_packed<Triangle1, StorageOrder>,
+  Triangle2>
+{
+  static constexpr bool value = !std::is_same_v<Triangle1, Triangle2>;
+};
+
+template <typename Layout, typename Triangle>
+inline constexpr bool triangle_layout_mismatch_v = triangle_layout_mismatch<Layout, Triangle>::value;
+
+};
+
+// Rank-1 update of a Symmetric matrix
+// performs BLAS xSYR/xSPR: A[i,j] += x[i] * x[j]
+
+template<class ExecSpace,
+         class ElementType_x,
+         std::experimental::extents<>::size_type ext_x,
+         class Layout_x,
+         class ElementType_A,
+         std::experimental::extents<>::size_type numRows_A,
+         std::experimental::extents<>::size_type numCols_A,
+         class Layout_A,
+         class Triangle>
+  requires (Impl::is_unique_layout_v<Layout_A, numRows_A, numCols_A>
+            or Impl::is_layout_blas_packed_v<Layout_A>)
+void symmetric_matrix_rank_1_update(kokkos_exec<ExecSpace> &&exec,
+  std::experimental::mdspan<ElementType_x, std::experimental::extents<ext_x>, Layout_x,
+    std::experimental::default_accessor<ElementType_x>> x,
+  std::experimental::mdspan<ElementType_A, std::experimental::extents<numRows_A, numCols_A>, Layout_A,
+    std::experimental::default_accessor<ElementType_A>> A,
+  Triangle t)
+{
+  // constraints
+  static_assert(A.rank() == 2);
+  static_assert(x.rank() == 1);
+  static_assert(!Impl::triangle_layout_mismatch_v<Layout_A, Triangle>);
+
+  // preconditions
+  if ( A.extent(0) != A.extent(1) ){
+    throw std::runtime_error("KokkosBlas: symmetric_matrix_rank_1_update: A.extent(0) != A.extent(1)");
+  }
+  if ( A.extent(0) != x.extent(0) ){
+    throw std::runtime_error("KokkosBlas: symmetric_matrix_rank_1_update: A.extent(0) != x.extent(0)");
+  }
+
+#if defined KOKKOS_STDBLAS_ENABLE_TESTS
+  std::cout << "symmetric_matrix_rank1_update: kokkos impl\n";
```

Comment:
Would you consider encapsulating the logger instead of using `<iostream>` objects directly?  For example:

```c++
extern void logKokkos(std::string_view functionName);
```

and then
```c++
#if defined(KOKKOS_STDBLAS_ENABLE_TESTS)
logKokkos("symmetric_matrix_rank1_update");
#endif
```